### PR TITLE
Upgrade conditions

### PR DESCRIPTION
**Pull request by ValClarkson**
**Checklist:**

 
 - [ ] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [ ] Have you tested your changes on all related environments with successful results, as applicable?
   - [ ] Have you added automated tests?

**Type of Changes:**

 
 - [x ] New feature
 - [ ] Bug fix
 - [ ] Documentation
 - [ ] Testing enhancement
 - [ ] Other

**What is the current behavior (link to any open issues here)?**

**What is the new behavior (if this is a feature change)?**
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Added conditions for upgrading the cluster

**Other Information**:

### CODE REVIEW

**Review by dsessler7 — CHANGES_REQUESTED** [review]
So I think we want something similar to what Ben did on lines 259-274 in `crunchybridgecluster_controller.go` where we check to see if an error condition is in place. If there is an error condition, we check the observed generation. If the observed generation hasn't changed, meaning the customer hasn't changed the spec, then we want to return early.

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 1]
```diff
@@ -350,6 +350,15 @@ func (r *CrunchyBridgeClusterReconciler) Reconcile(ctx context.Context, req ctrl
...
 		crunchybridgecluster.Status.ID = cluster.ID
+
+		meta.SetStatusCondition(&crunchybridgecluster.Status.Conditions, metav1.Condition{
+			ObservedGeneration: crunchybridgecluster.GetGeneration(),
+			Type:               v1beta1.ConditionUpgrading,
+			Status:             metav1.ConditionUnknown,
+			Reason:             "NoUpgradesInProgress",
+			Message: fmt.Sprintf(
+				"No upgrades in progress for Crunchy Bridge Cluster %v", crunchybridgecluster.Name),
+		})
```
[the remark this answers is not in the document]
Comment (ValClarkson) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go]: Updated

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 2]
```diff
@@ -400,6 +409,17 @@ func (r *CrunchyBridgeClusterReconciler) Reconcile(ctx context.Context, req ctrl
...
 	if len(crunchybridgecluster.Status.OngoingUpgrade) != 0 {
+		for _, operation := range clusterUpgradeDetails.Operations {
+			meta.SetStatusCondition(&crunchybridgecluster.Status.Conditions, metav1.Condition{
+				ObservedGeneration: crunchybridgecluster.GetGeneration(),
+				Type:               v1beta1.ConditionUpgrading,
+				Status:             metav1.ConditionTrue,
+				Reason:             operation.Flavor,
+				Message: fmt.Sprintf(
+					"Performing an upgrade of type %v with a state of %v on Crunchy Bridge Cluster %v",
+					operation.Flavor, operation.State, crunchybridgecluster.Name),
```
[the remark this answers is not in the document]
Comment (ValClarkson) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go]: removed everywhere I had it in the conditions

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 3]
```diff
@@ -479,11 +507,31 @@ func (r *CrunchyBridgeClusterReconciler) handleUpgrade(ctx context.Context,
...
 		// and what different results/requeue times we want to return.
 		// Currently: don't requeue and wait for user to change spec.
+		meta.SetStatusCondition(&crunchybridgecluster.Status.Conditions, metav1.Condition{
+			ObservedGeneration: crunchybridgecluster.GetGeneration(),
+			Type:               v1beta1.ConditionUpgrading,
+			Status:             metav1.ConditionFalse,
+			Reason:             "UpgradeError",
+			Message: fmt.Sprintf(
+				"Error performing an upgrade, please check your spec for errors or invalid values"+
+					"for cluster %v", crunchybridgecluster.Name),
```
[the remark this answers is not in the document]
Comment (ValClarkson) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go]: updated

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 4]
```diff
@@ -507,11 +555,28 @@ func (r *CrunchyBridgeClusterReconciler) handleUpgradeHA(ctx context.Context,
...
 		// and what different results/requeue times we want to return.
 		// Currently: don't requeue and wait for user to change spec.
+		meta.SetStatusCondition(&crunchybridgecluster.Status.Conditions, metav1.Condition{
+			ObservedGeneration: crunchybridgecluster.GetGeneration(),
+			Type:               v1beta1.ConditionUpgrading,
+			Status:             metav1.ConditionFalse,
+			Reason:             "HAUpgradeError",
+			Message: fmt.Sprintf(
+				"Error performing an HA upgrade, please check your spec for errors or invalid values"+
+					"for cluster %v", crunchybridgecluster.Name),
```
[the remark this answers is not in the document]
Comment (ValClarkson) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go]: updated

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 5]
```diff
@@ -507,11 +555,28 @@ func (r *CrunchyBridgeClusterReconciler) handleUpgradeHA(ctx context.Context,
...
 	}
 	clusterUpgrade.AddDataToClusterStatus(crunchybridgecluster)
 
+	meta.SetStatusCondition(&crunchybridgecluster.Status.Conditions, metav1.Condition{
+		ObservedGeneration: crunchybridgecluster.GetGeneration(),
+		Type:               v1beta1.ConditionUpgrading,
+		Status:             metav1.ConditionTrue,
+		Reason:             "UpgradeInProgress",
+		Message: fmt.Sprintf(
+			"HA upgrade in progress to %v on the Crunchy Bridge Cluster %v", action, crunchybridgecluster.Name),
```
[the remark this answers is not in the document]
Comment (ValClarkson) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go]: fixed

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 6]
```diff
@@ -479,11 +507,31 @@ func (r *CrunchyBridgeClusterReconciler) handleUpgrade(ctx context.Context,
...
+		meta.SetStatusCondition(&crunchybridgecluster.Status.Conditions, metav1.Condition{
+			ObservedGeneration: crunchybridgecluster.GetGeneration(),
+			Type:               v1beta1.ConditionUpgrading,
+			Status:             metav1.ConditionTrue,
+			Reason:             operation.Flavor,
+			Message: fmt.Sprintf(
+				"Performing an upgrade of type %v with a state of %v on Crunchy Bridge Cluster %v",
+				operation.Flavor, operation.State, crunchybridgecluster.Name),
+		})
+	}
```
[the remark this answers is not in the document]
Comment (ValClarkson) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go]: updated

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 7]
```diff
@@ -400,6 +409,17 @@ func (r *CrunchyBridgeClusterReconciler) Reconcile(ctx context.Context, req ctrl
 	// TODO(crunchybridgecluster): Do we want the operator to interrupt
 	// upgrades created through the GUI/API?
 	if len(crunchybridgecluster.Status.OngoingUpgrade) != 0 {
+		for _, operation := range clusterUpgradeDetails.Operations {
```
[the remark this answers is not in the document]
Comment (ValClarkson) [reply internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go]: updated

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 8]
```diff
@@ -265,12 +265,13 @@ func (r *CrunchyBridgeClusterReconciler) Reconcile(ctx context.Context, req ctrl
 		return ctrl.Result{}, nil
 	}
 
-	// Remove cluster invalid status if found
-	if invalid != nil &&
-		invalid.Status == metav1.ConditionFalse &&
-		invalid.Reason == "ClusterInvalid" {
-		meta.RemoveStatusCondition(&crunchybridgecluster.Status.Conditions,
-			v1beta1.ConditionCreating)
+	invalidUpgrade := meta.FindStatusCondition(crunchybridgecluster.Status.Conditions,
```
Comment (dsessler7) [inline internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go:268]: :pick: A comment above this block of code explaining what it is doing would be nice.

File: internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go [hunk 9]
```diff
@@ -507,11 +553,29 @@ func (r *CrunchyBridgeClusterReconciler) handleUpgradeHA(ctx context.Context,
 		// TODO(crunchybridgecluster): consider what errors we might get
 		// and what different results/requeue times we want to return.
 		// Currently: don't requeue and wait for user to change spec.
+		meta.SetStatusCondition(&crunchybridgecluster.Status.Conditions, metav1.Condition{
+			ObservedGeneration: crunchybridgecluster.GetGeneration(),
+			Type:               v1beta1.ConditionUpgrading,
+			Status:             metav1.ConditionFalse,
+			Reason:             "UpgradeError",
+			Message: fmt.Sprintf(
+				"Error performing an HA upgrade %v:", err),
```
Comment (dsessler7) [inline internal/bridge/crunchybridgecluster/crunchybridgecluster_controller.go:562]:
```suggestion
				"Error performing an HA upgrade: %v", err),
```

**Review by dsessler7 — APPROVED** [review]
Couple nitpicks, but looks good otherwise.

**Pull request merged**